File: XMLCheck/Handler/CheckRulesXML04/Xml4RuleBase.cs
using XMLCheck.Models;
using XMLCheck.Services;

namespace XMLCheck.Handler.CheckRulesXML04
{
    public abstract class Xml4RuleBase : IXml4Rule
    {
        public abstract string Key { get; }

        public abstract ErrorDetails? Check(Xml4Model model,HoSoContext context);

        protected ErrorDetails Error(string message)
        {
            return new ErrorDetails
            {
                key = Key,
                message = message
            };
        }
    }
}

File: XMLValidator_v2/src/parsers/rules/xml4/xml4RuleBase.ts
import { Xml4Model } from '../../../interfaces/xml-models.interface';
import { HoSoContext } from '../../../services/hoso-context';
import { ErrorDetails } from '../../../dto/xml-check.dto';

export interface IXml4Rule {
  readonly key: string;
  check(model: Xml4Model, context: HoSoContext): ErrorDetails | null;
}

export abstract class Xml4RuleBase implements IXml4Rule {
  public abstract get key(): string;
  public abstract check(model: Xml4Model, context: HoSoContext): ErrorDetails | null;

  protected error(message: string): ErrorDetails {
    return {
      key: this.key,
      message,
    };
  }
}
